Preliminary support for timers.

havardh opened the pull request.

havardh committed:
Preliminary support for timers.

havardh committed:
Removed private module and added constructor for Timer structure that initializes the struct with the memory mapped values.

havardh commented:
Seems like we have the reference to memory mapped values partially sovled. :)

I'm no entirely sure about the timer::Idx enum. But its a safe way to provide constructions for all the timers but parhaps we do not want the function to be called new either. This implies mem-allocation an this it does not allocate. 

This can be extended with:

```
impl Timer {
    pub fn new(idx: Idx) -> &'static Timer {
        unsafe {
            match idx {
                Idx::Timer0 => transmute(0x40010000),
                Idx::Timer1 => transmute(0x40010400),
                Idx::Timer2 => transmute(0x40010800),
                Idx::Timer3 => transmute(0x40010C00)
            }
        }
    }

    pub fn int_enable(&self, flags: u32) {
        unsafe { int_enable(transmute(self), flags) }
    }
}
```

Now we can do:

```
let timer0 = timer::Timer::new(timer::Idx::Timer0);
timer0.int_enable(TIMER_IF_OF);
```

havardh committed:
Added code to retreive the TIMER addresses from the C.

sondrele commented:
This is a lot better than what we had earlier.

But I don't really see why we need the `Idx` (Index?) enum. What about replacing the `new` function with the calls to get the different timers directly?

```
impl Timer {
    #[inline]
    pub fn timer0() -> &'static Timer {
        unsafe { transmute(GET_TIMER0()) }
    }

    #[inline]
    pub fn timer1() -> &'static Timer {
        unsafe { transmute(GET_TIMER1()) }
    }
    // ...
}

use timer::Timer;
let timer0 = Timer::timer0();
```

sondrele commented on emlib/timer.rs:
```diff
@@ -0,0 +1,442 @@
+}
+
+
+pub static TIMER_IF_OF: u32     = (0x1 << 0);
```
Also, we should be careful where we use `pub static` for identfiers, compared to `pub const`.
See: https://github.com/rust-lang/rfcs/blob/master/text/0246-const-vs-static.md

sondrele commented on emlib/timer.rs:
```diff
@@ -0,0 +1,442 @@
+pub static TIMER_IF_CC2: u32    = (0x1 << 6);
+pub static TIMER_IF_ICBOF0: u32 = (0x1 << 8);
+pub static TIMER_IF_ICBOF1: u32 = (0x1 << 9);
+pub static TIMER_IF_ICBOF2: u32 = (0x1 << 10);
```
Also, we should think about where we use `pub static` for identfiers, instead of `pub const`.
See: https://github.com/rust-lang/rfcs/blob/master/text/0246-const-vs-static.md

havardh replied:
Cool, that's what I want. Is this RFC implemented?

sondrele commented on emlib/timer.rs:
```diff
@@ -0,0 +1,442 @@
+    DTOGEN: u32,
+    DTFAULT: u32,
+    DTFAULTC: u32,
+    DTLOCK: u32,
```
We should consider defining all struct fields to `pub`. I think it makes more sense to have the fields public, at least for the bindings-library.

sondrele commented on emlib/timer.rs:
```diff
@@ -0,0 +1,442 @@
+#[derive(Copy)]
+pub enum Edge {
+    Rising = 0x0,
+    Faling = 0x1,
```
Rename to `Falling`.

havardh committed:
This changes the structure of the em_timer bindings to utilize traits.

- Makes fields in Timer struct public.
- Removes calls to transmute
- Makes timer parameter values constant.

havardh committed:
Moved the const variables to the top of the file.

havardh committed:
Added allow unstable to timer example

havardh committed:
Removed unused mut declaration

havardh committed:
Merge branch 'master' into emlib-timer

Conflicts:
	emlib/cmu.rs
	emlib/mod.rs
